# Data Preprocessing

## Data Acquisition

This is the "acquisition" of data for the country selector project. I do not intend to publish my work in hopes for any material benefits, however, I will list here all the data resources I have used for this project.

Firstly, I would like to take you through my thought process. At first, I wanted to use a Kaggle data set upon which I could try my hands at clustering. This would have been done through the [world happiness report dataset](https://www.kaggle.com/datasets/mathurinache/world-happiness-report). But I wanted to build it as a project that would be useful to many people in general. So I came up with the idea of a country selector. I have all these clusters of countries which have a lot of things in common; I check what those things are and I use them as "group features". These group features will then be presented to the user as a choice and then the user will see what group of countries scores the best in those areas, hence giving him an idea of what countries should he look more into.

All this having been said, I have done a fast (I still wanted to keep this a small project) research in order to see what are the most important criteria when choosing another place to live. I have documented this small endeavour in [this file](research.txt). Keeping these measures in mind, I browsed the web for data about countries' healthcare system, finance, climate, laws etc. and downloaded them. Again, keeping it short, since data was not available on one site, I just downloaded it instead of learning how to use different APIs. 

I have used the following websites to download my data and I give them full credit for this data:

https://climatedata.worldbank.org

https://worldpopulationreview.com

https://databank.worldbank.org

https://www.kaggle.com/datasets/mathurinache/world-happiness-report


## Data Preprocessing

This notebook will deal with transforming all the data obtained into one smooth dataset. Therefore here I will combine all these datasets into one. I will drop all the countries which have less than 70% of necessary data or (if possible) find a solution to impute missing data. The data is also not similar in terms of timeline. Therefore, if the data is available, I will select the most recent one. All the data I have downloaded is at most 4 years old (2018) at the moment of this notebook's creation.

I will also see if I can transform data values so that they are on similar scales, since it should (in theory) be better for clustering. I will apply other aggregations and replacement where I see fit.

Therefore there will be 4 steps, 1 for each type of data. The 5th an final step will be getting all these datasets cleaned and into one useful dataset to be used in the main part.

Without further ado, let's wrangle some data.

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
import re
import sys
# Insert at 1, 0 is the script path
sys.path.insert(1, './python-scripts/development')
from preprocessing_functions import categorize_gdp_per_capita_value, preprocess_world_bank_data, categorize_secure_internet_servers_per_million_people
import xarray as xr

**PNC:C:\Users\Andrei\anaconda3\lib\site-packages\PseudoNetCDF\pncwarn.py:24:UserWarning:
  pyproj could not be found, so IO/API coordinates cannot be converted to lat/lon; to fix, install pyproj or basemap (e.g., `pip install pyproj)`


## Part one - world happiness report

I will now see all the useful data I can extract from the world happiness report.

In [2]:
world_happiness = pd.read_csv("../data/world-happiness-2022.csv", index_col="RANK", decimal=",")
world_happiness.head()

Country  Happiness score  Whisker-high  Whisker-low  \
RANK                                                            
1         Finland            7.821         7.886        7.756   
2         Denmark            7.636         7.710        7.563   
3         Iceland            7.557         7.651        7.464   
4     Switzerland            7.512         7.586        7.437   
5     Netherlands            7.415         7.471        7.359   

      Dystopia (1.83) + residual  Explained by: GDP per capita  \
RANK                                                             
1                          2.518                         1.892   
2                          2.226                         1.953   
3                          2.320                         1.936   
4                          2.153                         2.026   
5                          2.137                         1.945   

      Explained by: Social support  Explained by: Healthy life expectancy  \
RANK                                                                        
1                            1.258                                  0.775   
2                            1.243                                  0.777   
3                            1.320                                  0.803   
4                            1.226                                  0.822   
5                            1.206                                  0.787   

      Explained by: Freedom to make life choices  Explained by: Generosity  \
RANK                                                                         
1                                          0.736                     0.109   
2                                          0.719                     0.188   
3                                          0.718                     0.270   
4                                          0.677                     0.147   
5                                          0.651                     0.271   

      Explained by: Perceptions of corruption  
RANK                                           
1                                       0.534  
2                                       0.532  
3                                       0.191  
4                                       0.461  
5                                       0.419

In [3]:
# See if there are any null values etc.
world_happiness.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 147 entries, 1 to 147
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     147 non-null    object 
 1   Happiness score                             146 non-null    float64
 2   Whisker-high                                146 non-null    float64
 3   Whisker-low                                 146 non-null    float64
 4   Dystopia (1.83) + residual                  146 non-null    float64
 5   Explained by: GDP per capita                146 non-null    float64
 6   Explained by: Social support                146 non-null    float64
 7   Explained by: Healthy life expectancy       146 non-null    float64
 8   Explained by: Freedom to make life choices  146 non-null    float64
 9   Explained by: Generosity                    146 non-null    float64
 10  Explained by: 

In [4]:
# Check entries with null values
world_happiness[pd.isnull(world_happiness).any(axis=1)]

Country  Happiness score  Whisker-high  Whisker-low  \
RANK                                                       
147       xx              NaN           NaN          NaN   

      Dystopia (1.83) + residual  Explained by: GDP per capita  \
RANK                                                             
147                          NaN                           NaN   

      Explained by: Social support  Explained by: Healthy life expectancy  \
RANK                                                                        
147                            NaN                                    NaN   

      Explained by: Freedom to make life choices  Explained by: Generosity  \
RANK                                                                         
147                                          NaN                       NaN   

      Explained by: Perceptions of corruption  
RANK                                           
147                                       NaN

In [5]:
# It looks like the 147th entry is not an actual country, so I can drop it.
world_happiness = world_happiness.drop(index=147)
world_happiness.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 146 entries, 1 to 146
Data columns (total 11 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Country                                     146 non-null    object 
 1   Happiness score                             146 non-null    float64
 2   Whisker-high                                146 non-null    float64
 3   Whisker-low                                 146 non-null    float64
 4   Dystopia (1.83) + residual                  146 non-null    float64
 5   Explained by: GDP per capita                146 non-null    float64
 6   Explained by: Social support                146 non-null    float64
 7   Explained by: Healthy life expectancy       146 non-null    float64
 8   Explained by: Freedom to make life choices  146 non-null    float64
 9   Explained by: Generosity                    146 non-null    float64
 10  Explained by: 

In [6]:
# I will also drop the whisker high, whisker low, and dystopia features as being irrelevant for the state of our project
world_happiness = world_happiness.drop(columns=["Whisker-high", "Whisker-low", "Dystopia (1.83) + residual"])
world_happiness.head()

Country  Happiness score  Explained by: GDP per capita  \
RANK                                                               
1         Finland            7.821                         1.892   
2         Denmark            7.636                         1.953   
3         Iceland            7.557                         1.936   
4     Switzerland            7.512                         2.026   
5     Netherlands            7.415                         1.945   

      Explained by: Social support  Explained by: Healthy life expectancy  \
RANK                                                                        
1                            1.258                                  0.775   
2                            1.243                                  0.777   
3                            1.320                                  0.803   
4                            1.226                                  0.822   
5                            1.206                                  0.787   

      Explained by: Freedom to make life choices  Explained by: Generosity  \
RANK                                                                         
1                                          0.736                     0.109   
2                                          0.719                     0.188   
3                                          0.718                     0.270   
4                                          0.677                     0.147   
5                                          0.651                     0.271   

      Explained by: Perceptions of corruption  
RANK                                           
1                                       0.534  
2                                       0.532  
3                                       0.191  
4                                       0.461  
5                                       0.419

In [7]:
# Furthermore, we will rename some of the features so we can use them more easily
world_happiness.rename(columns=lambda c: c.replace("Explained by: ", "").replace(" ", "_").lower(), inplace=True)
world_happiness.head()

country  happiness_score  gdp_per_capita  social_support  \
RANK                                                                 
1         Finland            7.821           1.892           1.258   
2         Denmark            7.636           1.953           1.243   
3         Iceland            7.557           1.936           1.320   
4     Switzerland            7.512           2.026           1.226   
5     Netherlands            7.415           1.945           1.206   

      healthy_life_expectancy  freedom_to_make_life_choices  generosity  \
RANK                                                                      
1                       0.775                         0.736       0.109   
2                       0.777                         0.719       0.188   
3                       0.803                         0.718       0.270   
4                       0.822                         0.677       0.147   
5                       0.787                         0.651       0.271   

      perceptions_of_corruption  
RANK                             
1                         0.534  
2                         0.532  
3                         0.191  
4                         0.461  
5                         0.419

In [8]:
# Another important step is scaling.
# I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups

# Get numerical columns for the world happiness dataset
numerical_columns_w_h = world_happiness.columns.copy().drop("country")

min_max_scaler = MinMaxScaler()
world_happiness[numerical_columns_w_h] = min_max_scaler.fit_transform(world_happiness[numerical_columns_w_h])
world_happiness.head()

country  happiness_score  gdp_per_capita  social_support  \
RANK                                                                 
1         Finland         1.000000        0.856496        0.953030   
2         Denmark         0.965848        0.884110        0.941667   
3         Iceland         0.951265        0.876415        1.000000   
4     Switzerland         0.942957        0.917157        0.928788   
5     Netherlands         0.925051        0.880489        0.913636   

      healthy_life_expectancy  freedom_to_make_life_choices  generosity  \
RANK                                                                      
1                    0.822718                      0.994595    0.232906   
2                    0.824841                      0.971622    0.401709   
3                    0.852442                      0.970270    0.576923   
4                    0.872611                      0.914865    0.314103   
5                    0.835456                      0.879730    0.579060   

      perceptions_of_corruption  
RANK                             
1                      0.909710  
2                      0.906303  
3                      0.325383  
4                      0.785349  
5                      0.713799

## Part 2 - databank datasets

This is the biggest chunk of the work. There are 10 datasets here, each of which has data for years 2020 and 2021 for every country. I will choose data from 2020 where it is not available for 2021. If no data is available, then data for that country will be null.

I will take all the datasets and try to compile them into one. For countries which miss most of the values I will drop them. For columns which do not have most of the values, I will drop them.

In [9]:
economy_stats = pd.read_excel("../data/Economy.xlsx", na_values="..")
economy_stats.head()

Country Name Country Code                                 Series Name  \
0  Afghanistan          AFG  Adjusted net national income (current US$)   
1  Afghanistan          AFG          Current account balance (% of GDP)   
2  Afghanistan          AFG                GDP per capita (current US$)   
3  Afghanistan          AFG          GNI per capita (constant 2015 US$)   
4      Albania          ALB  Adjusted net national income (current US$)   

         Series Code  2020 [YR2020]  2021 [YR2021]  
0     NY.ADJ.NNTY.CD   1.845879e+10            NaN  
1  BN.CAB.XOKA.GD.ZS  -1.559312e+01            NaN  
2     NY.GDP.PCAP.CD   5.167479e+02            NaN  
3     NY.GNP.PCAP.KD            NaN            NaN  
4     NY.ADJ.NNTY.CD   1.193938e+10            NaN

In [10]:
economy_stats = preprocess_world_bank_data(economy_stats)
economy_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 5 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   country_name                                195 non-null    object 
 1   Adjusted net national income (current US$)  170 non-null    float64
 2   Current account balance (% of GDP)          167 non-null    float64
 3   GDP per capita (current US$)                194 non-null    float64
 4   GNI per capita (constant 2015 US$)          128 non-null    float64
dtypes: float64(4), object(1)
memory usage: 7.7+ KB


1. It looks like for almost all countries I have the GDP per capita. I will check which country does not and through a quick google search see if I can find it.
2. Since all this is capital related, it makes sense that to fill the adjusted net national income and current account balance, I can group countries by GDP per capita and then fill missing values with the mean of the group.
3. Since for GNI per capita there are fewer than 70% of the values (65%), I will drop it.

In [11]:
# Check what kind of values we have so we can see what kind of groups we can create
economy_stats["GDP per capita (current US$)"].describe()

count       194.000000
mean      15348.067094
std       23405.739594
min         238.990726
25%        2169.762588
50%        5467.472829
75%       17714.185985
max      173688.189360
Name: GDP per capita (current US$), dtype: float64

In [12]:
# Drop GNI per capita column
economy_stats = economy_stats.drop(columns="GNI per capita (constant 2015 US$)")

# Check what country does not have the GDP per capita value assigned
economy_stats[economy_stats["GDP per capita (current US$)"].isnull()]

series_name country_name  Adjusted net national income (current US$)  \
162          South Sudan                                         NaN   

series_name  Current account balance (% of GDP)  GDP per capita (current US$)  
162                                  -35.007727                           NaN

In [13]:
# According to https://statisticstimes.com/economy/country/south-sudan-gdp-per-capita.php
# and https://knoema.com/atlas/South-Sudan/GDP-per-capita
# I can conclude that the GDP per capita for Sudan cane be set at 296
economy_stats.loc[162, "GDP per capita (current US$)"] = 296.0
economy_stats[economy_stats["GDP per capita (current US$)"].isnull()]

Empty DataFrame
Columns: [country_name, Adjusted net national income (current US$), Current account balance (% of GDP), GDP per capita (current US$)]
Index: []

In [14]:
# Get numerical columns for the economy stats dataset
numerical_columns_e_s = economy_stats.columns.copy().drop("country_name")

# Fill NA values with GDP per capita group mean
economy_stats[numerical_columns_e_s] = economy_stats.groupby(
    economy_stats["GDP per capita (current US$)"].apply(
        lambda x: categorize_gdp_per_capita_value)).transform(lambda x: x.fillna(x.mean()))
economy_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 4 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   country_name                                195 non-null    object 
 1   Adjusted net national income (current US$)  195 non-null    float64
 2   Current account balance (% of GDP)          195 non-null    float64
 3   GDP per capita (current US$)                195 non-null    float64
dtypes: float64(3), object(1)
memory usage: 6.2+ KB


C:\Users\Andrei\AppData\Local\Temp\ipykernel_17112\774526914.py:5: FutureWarning: Dropping invalid columns in DataFrameGroupBy.transform is deprecated. In a future version, a TypeError will be raised. Before calling .transform, select only columns which should be valid for the function.
  economy_stats[numerical_columns_e_s] = economy_stats.groupby(


In [15]:
# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

economy_stats[numerical_columns_e_s] = min_max_scaler.fit_transform(economy_stats[numerical_columns_e_s])
economy_stats.head()

series_name    country_name  Adjusted net national income (current US$)  \
0               Afghanistan                                    0.001026   
1                   Albania                                    0.000657   
2                   Algeria                                    0.006776   
3            American Samoa                                    0.021364   
4                    Angola                                    0.002009   

series_name  Current account balance (% of GDP)  GDP per capita (current US$)  
0                                      0.277733                      0.001601  
1                                      0.372249                      0.028868  
2                                      0.320042                      0.017687  
3                                      0.457369                      0.072678  
4                                      0.516524                      0.008862

In [16]:
education_stats = pd.read_excel("../data/education.xlsx", na_values="..")
education_stats.head()

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name        Series Code  \
0  Adjusted net enrollment rate, primary (% of pr...        SE.PRM.TENR   
1  Government expenditure on education, total (% ...  SE.XPD.TOTL.GD.ZS   
2  Literacy rate, adult total (% of people ages 1...     SE.ADT.LITR.ZS   
3  Educational attainment, at least Bachelor's or...  SE.TER.CUAT.BA.ZS   
4  Educational attainment, at least completed low...  SE.SEC.CUAT.LO.ZS   

   2020 [YR2020]  2021 [YR2021]  
0            NaN            NaN  
1            NaN            NaN  
2            NaN      37.266041  
3            NaN       3.067980  
4            NaN      11.631920

In [17]:
education_stats = preprocess_world_bank_data(education_stats)
education_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69 entries, 0 to 68
Data columns (total 5 columns):
 #   Column                                                                                              Non-Null Count  Dtype  
---  ------                                                                                              --------------  -----  
 0   country_name                                                                                        69 non-null     object 
 1   Educational attainment, at least Bachelor's or equivalent, population 25+, total (%) (cumulative)   38 non-null     float64
 2   Educational attainment, at least completed lower secondary, population 25+, total (%) (cumulative)  38 non-null     float64
 3   Government expenditure on education, total (% of GDP)                                               35 non-null     float64
 4   Literacy rate, adult total (% of people ages 15 and above)                                          16 non-null     flo

It looks like this data is mostly null, hence not useful at all. This means that I will need to find another source for education data. I will let this be for now. After processing all the available data I will see what kind of data I need and see if I can get it.

In [18]:
employment_stats = pd.read_excel("../data/employment.xlsx", na_values="..")
employment_stats.head()

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name            Series Code  \
0  Adequacy of social protection and labor progra...  per_allsp.adq_pop_tot   
1  Employers, total (% of total employment) (mode...         SL.EMP.MPYR.ZS   
2  Employment in industry (% of total employment)...         SL.IND.EMPL.ZS   
3  Employment in agriculture (% of total employme...         SL.AGR.EMPL.ZS   
4  Employment to population ratio, 15+, total (%)...   SL.EMP.TOTL.SP.NE.ZS   

   2020 [YR2020]  2021 [YR2021]  
0            NaN            NaN  
1            NaN            NaN  
2            NaN            NaN  
3            NaN            NaN  
4      36.709999            NaN

In [19]:
employment_stats = preprocess_world_bank_data(employment_stats)
employment_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99 entries, 0 to 98
Data columns (total 3 columns):
 #   Column                                                              Non-Null Count  Dtype  
---  ------                                                              --------------  -----  
 0   country_name                                                        99 non-null     object 
 1   Employment to population ratio, 15+, total (%) (national estimate)  95 non-null     float64
 2   Unemployment, total (% of total labor force) (national estimate)    99 non-null     float64
dtypes: float64(2), object(1)
memory usage: 2.4+ KB


It looks like only 2 columns were non null. I can use this data. Furthermore I can see that they mean more or less the same thing. Therefore I will keep the second statistic (unemployment rate).

I will scale it and then reverse it (take 1 - scaled_value) so that countries with low unemployment score low in this. I will do this because all of the statistics I have until now are positive functions, meaning that the bigger the value, the better the country in that particular area.

This is just my assumption that if we have the same measurements for each featue it will lead to better clustering.

In [20]:
employment_stats = employment_stats.drop(columns="Employment to population ratio, 15+, total (%) (national estimate)")
# Check if there are any more null values
employment_stats[employment_stats.isna().any(axis=1)]

Empty DataFrame
Columns: [country_name, Unemployment, total (% of total labor force) (national estimate)]
Index: []

In [21]:
# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

# Get numerical columns for the world happiness dataset
numerical_columns_em_s = employment_stats.columns.copy().drop("country_name")

employment_stats[numerical_columns_em_s] = min_max_scaler.fit_transform(employment_stats[numerical_columns_em_s])
employment_stats.head()

series_name country_name  \
0            Afghanistan   
1              Argentina   
2                Armenia   
3              Australia   
4                Austria   

series_name  Unemployment, total (% of total labor force) (national estimate)  
0                                                     0.397868                 
1                                                     0.389271                 
2                                                     0.620358                 
3                                                     0.170908                 
4                                                     0.179505

In [22]:
# Reverse statistic so high-value = good, low-value = bad; also rename column
employment_stats[numerical_columns_em_s] = 1.0 - employment_stats[numerical_columns_em_s]
employment_stats.rename(columns={'Unemployment, total (% of total labor force) (national estimate)': 'employment_rate_labor_force'})
employment_stats.head()

series_name country_name  \
0            Afghanistan   
1              Argentina   
2                Armenia   
3              Australia   
4                Austria   

series_name  Unemployment, total (% of total labor force) (national estimate)  
0                                                     0.602132                 
1                                                     0.610729                 
2                                                     0.379642                 
3                                                     0.829092                 
4                                                     0.820495

In [23]:
financial_indicators = pd.read_excel("../data/financial-indicators.xlsx", na_values="..")
financial_indicators.head(20)

Country Name Country Code  \
0      Afghanistan          AFG   
1      Afghanistan          AFG   
2      Afghanistan          AFG   
3      Afghanistan          AFG   
4          Albania          ALB   
5          Albania          ALB   
6          Albania          ALB   
7          Albania          ALB   
8          Algeria          DZA   
9          Algeria          DZA   
10         Algeria          DZA   
11         Algeria          DZA   
12  American Samoa          ASM   
13  American Samoa          ASM   
14  American Samoa          ASM   
15  American Samoa          ASM   
16         Andorra          AND   
17         Andorra          AND   
18         Andorra          AND   
19         Andorra          AND   

                                          Series Name     Series Code  \
0   Account ownership at a financial institution o...  FX.OWN.TOTL.ZS   
1                   Consumer price index (2010 = 100)     FP.CPI.TOTL   
2               Inflation, consumer prices (annual %)  FP.CPI.TOTL.ZG   
3                    Listed domestic companies, total  CM.MKT.LDOM.NO   
4   Account ownership at a financial institution o...  FX.OWN.TOTL.ZS   
5                   Consumer price index (2010 = 100)     FP.CPI.TOTL   
6               Inflation, consumer prices (annual %)  FP.CPI.TOTL.ZG   
7                    Listed domestic companies, total  CM.MKT.LDOM.NO   
8   Account ownership at a financial institution o...  FX.OWN.TOTL.ZS   
9                   Consumer price index (2010 = 100)     FP.CPI.TOTL   
10              Inflation, consumer prices (annual %)  FP.CPI.TOTL.ZG   
11                   Listed domestic companies, total  CM.MKT.LDOM.NO   
12  Account ownership at a financial institution o...  FX.OWN.TOTL.ZS   
13                  Consumer price index (2010 = 100)     FP.CPI.TOTL   
14              Inflation, consumer prices (annual %)  FP.CPI.TOTL.ZG   
15                   Listed domestic companies, total  CM.MKT.LDOM.NO   
16  Account ownership at a financial institution o...  FX.OWN.TOTL.ZS   
17                  Consumer price index (2010 = 100)     FP.CPI.TOTL   
18              Inflation, consumer prices (annual %)  FP.CPI.TOTL.ZG   
19                   Listed domestic companies, total  CM.MKT.LDOM.NO   

    2020 [YR2020]  2021 [YR2021]  
0             NaN            NaN  
1             NaN            NaN  
2             NaN            NaN  
3             NaN            NaN  
4             NaN            NaN  
5      120.978912     123.448662  
6        1.620887       2.041472  
7             NaN            NaN  
8             NaN            NaN  
9      155.017250     166.218895  
10       2.415131       7.226063  
11            NaN            NaN  
12            NaN            NaN  
13            NaN            NaN  
14            NaN            NaN  
15            NaN            NaN  
16            NaN            NaN  
17            NaN            NaN  
18            NaN            NaN  
19            NaN            NaN

In [24]:
financial_indicators = preprocess_world_bank_data(financial_indicators)
financial_indicators.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 164 entries, 0 to 163
Data columns (total 4 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   country_name                           164 non-null    object 
 1   Consumer price index (2010 = 100)      160 non-null    float64
 2   Inflation, consumer prices (annual %)  160 non-null    float64
 3   Listed domestic companies, total       69 non-null     float64
dtypes: float64(3), object(1)
memory usage: 5.2+ KB


I will drop 'listed domestic companies total'. As for the other 2 indexes, I will search which countries do not have it and will search for the values on the web.

Furthermore, after scaling, I will also revert the values (take the 1 - scaled_value) so that a value increase represents a good change.

In [25]:
financial_indicators.loc[(financial_indicators['Consumer price index (2010 = 100)'].isnull()) 
                         | (financial_indicators['Inflation, consumer prices (annual %)'].isnull())]

series_name        country_name  Consumer price index (2010 = 100)  \
10                     Barbados                                NaN   
27               Cayman Islands                                NaN   
68           Iran, Islamic Rep.                                NaN   
109                     Nigeria                                NaN   

series_name  Inflation, consumer prices (annual %)  \
10                                             NaN   
27                                             NaN   
68                                             NaN   
109                                            NaN   

series_name  Listed domestic companies, total  
10                                       15.0  
27                                        4.0  
68                                      367.0  
109                                     177.0

In [26]:
# Complete with values found on the web
financial_indicators.loc[10, 'Consumer price index (2010 = 100)'] = 212.3
financial_indicators.loc[10, 'Inflation, consumer prices (annual %)'] = 4.1003

financial_indicators.loc[27, 'Consumer price index (2010 = 100)'] = 121
financial_indicators.loc[27, 'Inflation, consumer prices (annual %)'] = -0.6

financial_indicators.loc[68, 'Consumer price index (2010 = 100)'] = 415
financial_indicators.loc[68, 'Inflation, consumer prices (annual %)'] = 39.9

financial_indicators.loc[109, 'Consumer price index (2010 = 100)'] = 439
financial_indicators.loc[109, 'Inflation, consumer prices (annual %)'] = 11.4

In [27]:
# Drop unuseful column
financial_indicators = financial_indicators.drop(columns="Listed domestic companies, total")

# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

# Get numerical columns for the world happiness dataset
numerical_columns_f_i = financial_indicators.columns.copy().drop("country_name")

financial_indicators[numerical_columns_f_i] = min_max_scaler.fit_transform(financial_indicators[numerical_columns_f_i])
financial_indicators.head()

series_name         country_name  Consumer price index (2010 = 100)  \
0                        Albania                           0.001183   
1                        Algeria                           0.003288   
2            Antigua and Barbuda                           0.000943   
3                        Armenia                           0.002004   
4                      Australia                           0.001224   

series_name  Inflation, consumer prices (annual %)  
0                                         0.007368  
1                                         0.016638  
2                                         0.007407  
3                                         0.016565  
4                                         0.008839

In [28]:
# Reverse statistic so high-value = good, low-value = bad; also rename column
financial_indicators[numerical_columns_f_i] = 1.0 - financial_indicators[numerical_columns_f_i]
financial_indicators.rename(columns={'Consumer price index (2010 = 100)': 'inverse_of_consumer_price_', 
                                     'Inflation, consumer prices (annual %)': 'inverse_of_inflation'})
financial_indicators.head()

series_name         country_name  Consumer price index (2010 = 100)  \
0                        Albania                           0.998817   
1                        Algeria                           0.996712   
2            Antigua and Barbuda                           0.999057   
3                        Armenia                           0.997996   
4                      Australia                           0.998776   

series_name  Inflation, consumer prices (annual %)  
0                                         0.992632  
1                                         0.983362  
2                                         0.992593  
3                                         0.983435  
4                                         0.991161

In [29]:
health_stats = pd.read_excel("../data/health.xlsx", na_values="..")
health_stats.head()

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name        Series Code  \
0  Current health expenditure per capita, PPP (cu...  SH.XPD.CHEX.PP.CD   
1                   Hospital beds (per 1,000 people)     SH.MED.BEDS.ZS   
2  Mortality caused by road traffic injury (per 1...     SH.STA.TRAF.P5   
3  Out-of-pocket expenditure per capita (current ...  SH.XPD.OOPC.PC.CD   
4  People with basic handwashing facilities inclu...     SH.STA.HYGN.ZS   

   2020 [YR2020]  2021 [YR2021]  
0            NaN            NaN  
1            NaN            NaN  
2            NaN            NaN  
3            NaN            NaN  
4       38.11505            NaN

In [30]:
health_stats = preprocess_world_bank_data(health_stats)
health_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 216 entries, 0 to 215
Data columns (total 3 columns):
 #   Column                                                                               Non-Null Count  Dtype  
---  ------                                                                               --------------  -----  
 0   country_name                                                                         216 non-null    object 
 1   People with basic handwashing facilities including soap and water (% of population)  79 non-null     float64
 2   Population growth (annual %)                                                         216 non-null    float64
dtypes: float64(2), object(1)
memory usage: 5.2+ KB


It looks like we can use the 'population growth' value, but not the other one. Hence, we will drop the unuseful one and keep the useful one.

In [31]:
# Drop columns with too many null values
health_stats = health_stats.drop(
    columns="People with basic handwashing facilities including soap and water (% of population)")

# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

# Get numerical columns for the world happiness dataset
numerical_columns_h_s = health_stats.columns.copy().drop("country_name")

health_stats[numerical_columns_h_s] = min_max_scaler.fit_transform(health_stats[numerical_columns_h_s])
health_stats.head()

series_name    country_name  Population growth (annual %)
0               Afghanistan                      0.730947
1                   Albania                      0.203302
2                   Algeria                      0.645278
3            American Samoa                      0.270417
4                   Andorra                      0.336833

In [32]:
infrastructure_stats = pd.read_excel("../data/infrastructure.xlsx", na_values="..")
infrastructure_stats.head()

Country Name Country Code                                      Series Name  \
0  Afghanistan          AFG   Fixed broadband subscriptions (per 100 people)   
1  Afghanistan          AFG                      Rail lines (total route-km)   
2  Afghanistan          AFG   Secure Internet servers (per 1 million people)   
3  Afghanistan          AFG  Research and development expenditure (% of GDP)   
4      Albania          ALB   Fixed broadband subscriptions (per 100 people)   

         Series Code  2020 [YR2020]  2021 [YR2021]  
0     IT.NET.BBND.P2       0.068254            NaN  
1     IS.RRS.TOTL.KM            NaN            NaN  
2     IT.NET.SECR.P6      34.987363            NaN  
3  GB.XPD.RSDV.GD.ZS            NaN            NaN  
4     IT.NET.BBND.P2      17.684951            NaN

In [33]:
infrastructure_stats = preprocess_world_bank_data(infrastructure_stats)
infrastructure_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 3 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   country_name                                    215 non-null    object 
 1   Fixed broadband subscriptions (per 100 people)  201 non-null    float64
 2   Secure Internet servers (per 1 million people)  214 non-null    float64
dtypes: float64(2), object(1)
memory usage: 5.2+ KB


It looks like we have 2 infrastructure indicators that we can use. For the column with the missing value I will search manually for the value, but for the one with 14 missing value, I will fill null values based on group means.

In [34]:
infrastructure_stats.loc[infrastructure_stats["Secure Internet servers (per 1 million people)"].isnull()]

series_name      country_name  Fixed broadband subscriptions (per 100 people)  \
150          Papua New Guinea                                        0.234715   

series_name  Secure Internet servers (per 1 million people)  
150                                                     NaN

In [35]:
# Fill with found value
infrastructure_stats.loc[150, "Secure Internet servers (per 1 million people)"] = 56.74
infrastructure_stats.loc[infrastructure_stats["Secure Internet servers (per 1 million people)"].isnull()]

Empty DataFrame
Columns: [country_name, Fixed broadband subscriptions (per 100 people), Secure Internet servers (per 1 million people)]
Index: []

In [36]:
# Get numerical columns for the economy stats dataset
numerical_columns_i_s = infrastructure_stats.columns.copy().drop("country_name")

# Fill NA values with GDP per capita group mean
infrastructure_stats[numerical_columns_i_s] = infrastructure_stats.groupby(
    infrastructure_stats["Fixed broadband subscriptions (per 100 people)"].apply(
        lambda x: categorize_secure_internet_servers_per_million_people)).transform(lambda x: x.fillna(x.mean()))
infrastructure_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 3 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   country_name                                    215 non-null    object 
 1   Fixed broadband subscriptions (per 100 people)  215 non-null    float64
 2   Secure Internet servers (per 1 million people)  215 non-null    float64
dtypes: float64(2), object(1)
memory usage: 5.2+ KB


C:\Users\Andrei\AppData\Local\Temp\ipykernel_17112\1489755229.py:5: FutureWarning: Dropping invalid columns in DataFrameGroupBy.transform is deprecated. In a future version, a TypeError will be raised. Before calling .transform, select only columns which should be valid for the function.
  infrastructure_stats[numerical_columns_i_s] = infrastructure_stats.groupby(


In [37]:
# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

infrastructure_stats[numerical_columns_i_s] = min_max_scaler.fit_transform(infrastructure_stats[numerical_columns_i_s])
infrastructure_stats.head()

series_name    country_name  Fixed broadband subscriptions (per 100 people)  \
0               Afghanistan                                        0.001095   
1                   Albania                                        0.283604   
2                   Algeria                                        0.138618   
3            American Samoa                                        0.265331   
4                   Andorra                                        0.767940   

series_name  Secure Internet servers (per 1 million people)  
0                                                  0.000047  
1                                                  0.001194  
2                                                  0.000065  
3                                                  0.000293  
4                                                  0.013116

In [38]:
population_and_environment_stats = pd.read_excel("../data/population-and-environment.xlsx", na_values="..")
population_and_environment_stats.head()

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name     Series Code  \
0            Access to electricity (% of population)  EG.ELC.ACCS.ZS   
1             CO2 emissions (metric tons per capita)  EN.ATM.CO2E.PC   
2  Electricity production from coal sources (% of...  EG.ELC.COAL.ZS   
3                                 Land area (sq. km)  AG.LND.TOTL.K2   
4                       Forest area (% of land area)  AG.LND.FRST.ZS   

   2020 [YR2020]  2021 [YR2021]  
0      97.699997            NaN  
1            NaN            NaN  
2            NaN            NaN  
3  652860.000000       652860.0  
4       1.850994            NaN

In [39]:
population_and_environment_stats = preprocess_world_bank_data(population_and_environment_stats)
population_and_environment_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 216 entries, 0 to 215
Data columns (total 6 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   country_name                                         216 non-null    object 
 1   Access to electricity (% of population)              215 non-null    float64
 2   Forest area (% of land area)                         210 non-null    float64
 3   Land area (sq. km)                                   216 non-null    float64
 4   Population density (people per sq. km of land area)  216 non-null    float64
 5   Urban population (% of total population)             214 non-null    float64
dtypes: float64(5), object(1)
memory usage: 10.2+ KB


Because of the low number of null values we can impute them value by value

In [40]:
population_and_environment_stats[population_and_environment_stats['Access to electricity (% of population)'].isnull()]

series_name    country_name  Access to electricity (% of population)  \
3            American Samoa                                      NaN   

series_name  Forest area (% of land area)  Land area (sq. km)  \
3                                   85.65               200.0   

series_name  Population density (people per sq. km of land area)  \
3                                                      275.985     

series_name  Urban population (% of total population)  
3                                              87.153

In [41]:
# Impute with found value on the web
population_and_environment_stats.loc[3, 'Access to electricity (% of population)'] = 99.2
population_and_environment_stats[population_and_environment_stats['Access to electricity (% of population)'].isnull()]

Empty DataFrame
Columns: [country_name, Access to electricity (% of population), Forest area (% of land area), Land area (sq. km), Population density (people per sq. km of land area), Urban population (% of total population)]
Index: []

In [42]:
population_and_environment_stats[population_and_environment_stats['Forest area (% of land area)'].isnull()]

series_name          country_name  Access to electricity (% of population)  \
39                Channel Islands                                    100.0   
75                      Gibraltar                                    100.0   
86           Hong Kong SAR, China                                    100.0   
116              Macao SAR, China                                    100.0   
129                        Monaco                                    100.0   
136                         Nauru                                    100.0   

series_name  Forest area (% of land area)  Land area (sq. km)  \
39                                    NaN             198.000   
75                                    NaN              10.000   
86                                    NaN            1050.000   
116                                   NaN              32.900   
129                                   NaN               2.027   
136                                   NaN              20.000   

series_name  Population density (people per sq. km of land area)  \
39                                                  878.075758     
75                                                 3369.100000     
86                                                 7125.523810     
116                                               19736.838906     
129                                               19360.631475     
136                                                 541.700000     

series_name  Urban population (% of total population)  
39                                             30.963  
75                                            100.000  
86                                            100.000  
116                                           100.000  
129                                           100.000  
136                                           100.000

In [43]:
# Impute with found values on the web
population_and_environment_stats.loc[39, 'Forest area (% of land area)'] = 5.15
population_and_environment_stats.loc[75, 'Forest area (% of land area)'] = 0
population_and_environment_stats.loc[86, 'Forest area (% of land area)'] = 56
population_and_environment_stats.loc[116, 'Forest area (% of land area)'] = 1.5
population_and_environment_stats.loc[129, 'Forest area (% of land area)'] = 0
population_and_environment_stats.loc[136, 'Forest area (% of land area)'] = 0

population_and_environment_stats[population_and_environment_stats['Forest area (% of land area)'].isnull()]

Empty DataFrame
Columns: [country_name, Access to electricity (% of population), Forest area (% of land area), Land area (sq. km), Population density (people per sq. km of land area), Urban population (% of total population)]
Index: []

In [44]:
population_and_environment_stats[population_and_environment_stats['Urban population (% of total population)'].isnull()]

series_name              country_name  \
61                            Eritrea   
182          St. Martin (French part)   

series_name  Access to electricity (% of population)  \
61                                         52.171097   
182                                       100.000000   

series_name  Forest area (% of land area)  Land area (sq. km)  \
61                              10.448119            101000.0   
182                             24.800000                50.0   

series_name  Population density (people per sq. km of land area)  \
61                                                   35.113139     
182                                                 773.180000     

series_name  Urban population (% of total population)  
61                                                NaN  
182                                               NaN

In [45]:
# Impute with found values on the web
population_and_environment_stats.loc[61, 'Urban population (% of total population)'] = 40.71
population_and_environment_stats.loc[182, 'Urban population (% of total population)'] = 0

population_and_environment_stats[population_and_environment_stats['Urban population (% of total population)'].isnull()]

Empty DataFrame
Columns: [country_name, Access to electricity (% of population), Forest area (% of land area), Land area (sq. km), Population density (people per sq. km of land area), Urban population (% of total population)]
Index: []

In [46]:
# Check if there are any more null values
population_and_environment_stats[population_and_environment_stats.isna().any(axis=1)]

Empty DataFrame
Columns: [country_name, Access to electricity (% of population), Forest area (% of land area), Land area (sq. km), Population density (people per sq. km of land area), Urban population (% of total population)]
Index: []

In [47]:
# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

# Get numerical columns for the world happiness dataset
numerical_columns_p_e_s = population_and_environment_stats.columns.copy().drop("country_name")

population_and_environment_stats[numerical_columns_p_e_s] = min_max_scaler.fit_transform(
    population_and_environment_stats[numerical_columns_p_e_s])
population_and_environment_stats.head()

series_name    country_name  Access to electricity (% of population)  \
0               Afghanistan                                 0.975204   
1                   Albania                                 1.000000   
2                   Algeria                                 0.997888   
3            American Samoa                                 0.991375   
4                   Andorra                                 1.000000   

series_name  Forest area (% of land area)  Land area (sq. km)  \
0                                0.019002            0.039865   
1                                0.295569            0.001673   
2                                0.008400            0.145433   
3                                0.879254            0.000012   
4                                0.349469            0.000029   

series_name  Population density (people per sq. km of land area)  \
0                                                     0.003014     
1                                                     0.005241     
2                                                     0.000926     
3                                                     0.013976     
4                                                     0.008322     

series_name  Urban population (% of total population)  
0                                             0.26026  
1                                             0.62112  
2                                             0.73733  
3                                             0.87153  
4                                             0.87916

In [48]:
poverty_stats = pd.read_excel("../data/poverty.xlsx", na_values="..")
poverty_stats.head()

Country Name Country Code  \
0     Afghanistan          AFG   
1         Albania          ALB   
2         Algeria          DZA   
3  American Samoa          ASM   
4         Andorra          AND   

                                         Series Name  Series Code  \
0  Multidimensional poverty headcount ratio (% of...  SI.POV.MDIM   
1  Multidimensional poverty headcount ratio (% of...  SI.POV.MDIM   
2  Multidimensional poverty headcount ratio (% of...  SI.POV.MDIM   
3  Multidimensional poverty headcount ratio (% of...  SI.POV.MDIM   
4  Multidimensional poverty headcount ratio (% of...  SI.POV.MDIM   

   2020 [YR2020]  2021 [YR2021]  
0           49.4            NaN  
1           43.4            NaN  
2            NaN            NaN  
3            NaN            NaN  
4            NaN            NaN

In [49]:
poverty_stats = preprocess_world_bank_data(poverty_stats)
poverty_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 2 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   country_name                                                      35 non-null     object 
 1   Multidimensional poverty headcount ratio (% of total population)  35 non-null     float64
dtypes: float64(1), object(1)
memory usage: 688.0+ bytes


Again, insufficient data, because there are only 35 countries with this data. Will do the same as stated above.

In [50]:
private_sector_stats = pd.read_excel("../data/private-sector.xlsx", na_values="..")
private_sector_stats.head()

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3      Albania          ALB   
4      Albania          ALB   

                                         Series Name        Series Code  \
0  Cost of business start-up procedures (% of GNI...  IC.REG.COST.PC.ZS   
1  Ease of doing business score (0 = lowest perfo...     IC.BUS.DFRN.XQ   
2  Labor tax and contributions (% of commercial p...  IC.TAX.LABR.CP.ZS   
3  Cost of business start-up procedures (% of GNI...  IC.REG.COST.PC.ZS   
4  Ease of doing business score (0 = lowest perfo...     IC.BUS.DFRN.XQ   

   2020 [YR2020]  2021 [YR2021]  
0            NaN            NaN  
1            NaN            NaN  
2            NaN            NaN  
3            NaN            NaN  
4            NaN            NaN

In [51]:
private_sector_stats = preprocess_world_bank_data(private_sector_stats)
private_sector_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   country_name  0 non-null      object
dtypes: object(1)
memory usage: 124.0+ bytes


Again, insufficient data. Will do the same as stated above.

In [52]:
public_sector_stats = pd.read_excel("../data/public-sector.xlsx", na_values="..")
public_sector_stats.head()

Country Name Country Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series Name     Series Code  \
0     CPIA financial sector rating (1=low to 6=high)  IQ.CPA.FINS.XQ   
1      CPIA gender equality rating (1=low to 6=high)  IQ.CPA.GNDR.XQ   
2  CPIA policies for social inclusion/equity clus...  IQ.CPA.SOCI.XQ   
3  CPIA property rights and rule-based governance...  IQ.CPA.PROP.XQ   
4  CPIA quality of public administration rating (...  IQ.CPA.PADM.XQ   

   2020 [YR2020]  2021 [YR2021]  
0            1.5            NaN  
1            1.5            NaN  
2            2.7            NaN  
3            2.0            NaN  
4            2.5            NaN

In [53]:
public_sector_stats = preprocess_world_bank_data(public_sector_stats)
public_sector_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73 entries, 0 to 72
Data columns (total 6 columns):
 #   Column                                                                       Non-Null Count  Dtype  
---  ------                                                                       --------------  -----  
 0   country_name                                                                 73 non-null     object 
 1   CPIA financial sector rating (1=low to 6=high)                               73 non-null     float64
 2   CPIA gender equality rating (1=low to 6=high)                                73 non-null     float64
 3   CPIA policies for social inclusion/equity cluster average (1=low to 6=high)  73 non-null     float64
 4   CPIA property rights and rule-based governance rating (1=low to 6=high)      73 non-null     float64
 5   CPIA quality of public administration rating (1=low to 6=high)               73 non-null     float64
dtypes: float64(5), object(1)
memory usage: 3.5+ K

In [54]:
# Rename new columns
public_sector_stats.rename(columns=lambda c: re.sub(r' \(.*\)', '', c), inplace=True)
public_sector_stats.rename(columns=lambda c: re.sub(r'CPIA ', '', c), inplace=True)
public_sector_stats.rename(columns=lambda c: c.replace(" ", "_").lower(), inplace=True)
public_sector_stats.head()

series_name  country_name  financial_sector_rating  gender_equality_rating  \
0             Afghanistan                      1.5                     1.5   
1              Bangladesh                      2.5                     3.0   
2                   Benin                      2.5                     3.5   
3                  Bhutan                      3.0                     4.0   
4            Burkina Faso                      3.0                     3.5   

series_name  policies_for_social_inclusion/equity_cluster_average  \
0                                                          2.7      
1                                                          3.3      
2                                                          3.5      
3                                                          4.0      
4                                                          3.6      

series_name  property_rights_and_rule-based_governance_rating  \
0                                                         2.0   
1                                                         2.5   
2                                                         3.5   
3                                                         4.0   
4                                                         3.0   

series_name  quality_of_public_administration_rating  
0                                                2.5  
1                                                2.0  
2                                                3.0  
3                                                4.0  
4                                                3.0

In [55]:
# Again, I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

# Get numerical columns for the world happiness dataset
numerical_columns_pu_s_s = public_sector_stats.columns.copy().drop("country_name")

public_sector_stats[numerical_columns_pu_s_s] = min_max_scaler.fit_transform(public_sector_stats[numerical_columns_pu_s_s])
public_sector_stats.head()

series_name  country_name  financial_sector_rating  gender_equality_rating  \
0             Afghanistan                 0.166667                0.000000   
1              Bangladesh                 0.500000                0.500000   
2                   Benin                 0.500000                0.666667   
3                  Bhutan                 0.666667                0.833333   
4            Burkina Faso                 0.666667                0.666667   

series_name  policies_for_social_inclusion/equity_cluster_average  \
0                                                     0.428571      
1                                                     0.642857      
2                                                     0.714286      
3                                                     0.892857      
4                                                     0.750000      

series_name  property_rights_and_rule-based_governance_rating  \
0                                                    0.333333   
1                                                    0.500000   
2                                                    0.833333   
3                                                    1.000000   
4                                                    0.666667   

series_name  quality_of_public_administration_rating  
0                                           0.500000  
1                                           0.333333  
2                                           0.666667  
3                                           1.000000  
4                                           0.666667

Up until this point we have 7 data sets which yielded results: economy_stats, public_sector_stats, financial_indicators, health_stats and population_and_environment_stats, employment_stats, infrastructure_stats.

The other 3 data sets did not yield any useful result: education_stats, poverty_stats, private_sector_stats.

I will search for this other type of data and see if I can find it quickly. If I cannot, I will leave it be and work with the available data.

## Part 3 - climate data sets

This consists of 2 files: precipitation and mean temperature.

In [61]:
precipitation_ds = xr.open_dataset('../data/precipitation.nc')
precipitation_stats = precipitation_ds.to_dataframe()
precipitation_stats.head(20)

climatology-pr-annual-mean  lon_bnds  lat_bnds
time       lat    lon     bnds                                                
1991-01-16 -89.75 -179.75 0                            NaN    -180.0     -90.0
                          1                            NaN    -179.5     -89.5
                  -179.25 0                            NaN    -179.5     -90.0
                          1                            NaN    -179.0     -89.5
                  -178.75 0                            NaN    -179.0     -90.0
                          1                            NaN    -178.5     -89.5
                  -178.25 0                            NaN    -178.5     -90.0
                          1                            NaN    -178.0     -89.5
                  -177.75 0                            NaN    -178.0     -90.0
                          1                            NaN    -177.5     -89.5
                  -177.25 0                            NaN    -177.5     -90.0
                          1                            NaN    -177.0     -89.5
                  -176.75 0                            NaN    -177.0     -90.0
                          1                            NaN    -176.5     -89.5
                  -176.25 0                            NaN    -176.5     -90.0
                          1                            NaN    -176.0     -89.5
                  -175.75 0                            NaN    -176.0     -90.0
                          1                            NaN    -175.5     -89.5
                  -175.25 0                            NaN    -175.5     -90.0
                          1                            NaN    -175.0     -89.5

It looks like this data is based on latitude and not on countries. It will be the same for temperature as I have downloaded from the same website. This is useless.

Therefore I decided to go on without it.

After a brief search on the web so get this data, I could not find something quick and free. because I already have some good info and I want to keep this project short, I will not proceeed any further. There will be no information about precipitation and temperature.

## Part 4 - crime statistics

This consists of 1 file: crime-rates.json

In [71]:
crime_rates = pd.read_json("../data/crime-rates.json")
crime_rates.head(10)

ranking              country  crimeIndex     pop2022
0        1            Venezuela       83.76   29266.991
1        2     Papua New Guinea       80.79    9292.169
2        3         South Africa       76.86   60756.135
3        4          Afghanistan       76.31   40754.388
4        5             Honduras       74.54   10221.247
5        6  Trinidad And Tobago       71.63    1406.585
6        7               Guyana       68.74     794.045
7        8          El Salvador       67.79    6550.389
8        9               Brazil       67.49  215353.593
9       10              Jamaica       67.42    2985.094

In [72]:
# Drop useless columns ranking and pop2022
crime_rates = crime_rates.drop(columns=['ranking', 'pop2022'])

# I will use a simple min max scaling so that all features are between 0 and 1 and they keep their impact 
# respective to their groups
min_max_scaler = MinMaxScaler()

# Get numerical columns for the world happiness dataset

crime_rates[['crimeIndex']] = min_max_scaler.fit_transform(crime_rates[['crimeIndex']])
crime_rates.head()

country  crimeIndex
0         Venezuela    1.000000
1  Papua New Guinea    0.958537
2      South Africa    0.903672
3       Afghanistan    0.895993
4          Honduras    0.871283

In [73]:
# Reverse statistic so high-value = good, low-value = bad; also rename column
crime_rates['crimeIndex'] = 1.0 - crime_rates['crimeIndex']
crime_rates.rename(columns={'crimeIndex': 'opposite_of_crime_index'}, inplace=True)
crime_rates.head()

country  opposite_of_crime_index
0         Venezuela             1.110223e-16
1  Papua New Guinea             4.146307e-02
2      South Africa             9.632835e-02
3       Afghanistan             1.040067e-01
4          Honduras             1.287170e-01

## Step 5 - putting them all together

This is the most important stept. We now have 1 + 7 + 0 + 1 = 9 data sets, each having different statistics whose values are from 0 = bad to 1 = great.

Each data set has statistics for a certain number of countries. There is a total of 195 countries in the world (according to Google). I would like to be able to include at least 70% of them in this project. Therefore:

- I will drop any data set with less than 136 countries
- If I consider I will have too few features for the 70% margin, I will drop it to 60%

I will see what are those countries that have statistics in all data sets and then build one big dataframe having statistics about all these countries. I will save this dataframe as a csv file with the name preprocessed_country_data.csv

This will be the starting point for the clustering project.

In [75]:
print(world_happiness.shape, economy_stats.shape, public_sector_stats.shape, financial_indicators.shape, 
      health_stats.shape, population_and_environment_stats.shape, employment_stats.shape, infrastructure_stats.shape,
      crime_rates.shape)

(146, 8) (195, 4) (73, 6) (164, 3) (216, 2) (216, 6) (99, 2) (215, 3) (136, 2)


It looks like we will need to lose the public_sector_stats as it is comprised of only 73 countries and also the employment_stats as it is comprised of only 99 countries.

I will therefore use the remaining 7 data sets and see if I can get them to agree on 136 countries (the lowest number of countries between these data sets).

I will take as reference these countries and then compare them with countries in all the other datasets. After seeing the difference in all, I will see what will be the remaining countries I will go with.

In [76]:
reference_countries = set(crime_rates['country'].values)
reference_countries

{'Afghanistan',
 'Albania',
 'Algeria',
 'Angola',
 'Argentina',
 'Armenia',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas',
 'Bahrain',
 'Bangladesh',
 'Belarus',
 'Belgium',
 'Belize',
 'Bolivia',
 'Bosnia And Herzegovina',
 'Botswana',
 'Brazil',
 'Brunei',
 'Bulgaria',
 'Cambodia',
 'Cameroon',
 'Canada',
 'Chile',
 'China',
 'Colombia',
 'Costa Rica',
 'Croatia',
 'Cuba',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Dominican Republic',
 'Ecuador',
 'Egypt',
 'El Salvador',
 'Estonia',
 'Ethiopia',
 'Fiji',
 'Finland',
 'France',
 'Georgia',
 'Germany',
 'Ghana',
 'Greece',
 'Guatemala',
 'Guyana',
 'Honduras',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Iran',
 'Iraq',
 'Ireland',
 'Isle Of Man',
 'Israel',
 'Italy',
 'Jamaica',
 'Japan',
 'Jordan',
 'Kazakhstan',
 'Kenya',
 'Kuwait',
 'Kyrgyzstan',
 'Latvia',
 'Lebanon',
 'Libya',
 'Lithuania',
 'Luxembourg',
 'Malaysia',
 'Maldives',
 'Malta',
 'Mauritius',
 'Mexico',
 'Moldova',
 'Mongolia',
 'Montenegr

In [80]:
# Difference between reference and world happiness
w_h_countries = set(world_happiness['country'].values)
print(reference_countries.difference(w_h_countries))
print(w_h_countries.difference(reference_countries))

{'Kuwait', 'Taiwan', 'Qatar', 'Maldives', 'Guyana', 'Somalia', 'Botswana', 'Hong Kong', 'Oman', 'Guatemala', 'Angola', 'Bahamas', 'Brunei', 'Isle Of Man', 'Syria', 'Libya', 'Sudan', 'Czech Republic', 'Papua New Guinea', 'Bosnia And Herzegovina', 'Palestine', 'Azerbaijan', 'Cuba', 'Belize', 'Luxembourg', 'Belarus', 'Trinidad And Tobago', 'Rwanda', 'Fiji', 'Puerto Rico'}
{'Turkmenistan*', 'Mozambique', 'Mauritania*', 'Guinea', 'Belarus*', 'North Cyprus*', 'Azerbaijan*', 'Sierra Leone', 'Czechia', 'Bosnia and Herzegovina', 'Laos', 'Congo', 'Comoros*', 'Togo', 'Benin', 'Rwanda*', 'Lesotho*', 'Palestinian Territories*', 'Hong Kong S.A.R. of China', 'Guatemala*', 'Yemen*', 'Kuwait*', 'Eswatini, Kingdom of*', 'Senegal', 'Luxembourg*', 'Kosovo', 'Mali', 'Ivory Coast', 'Malawi', 'Botswana*', 'Tajikistan', 'Gabon', 'Liberia*', 'Taiwan Province of China', 'Burkina Faso', 'Gambia*', 'Madagascar*', 'Chad*', 'Niger*', 'Libya*'}


I can see that there are lots of countries' names suffixed with a star in the world happiness dataset. I will strip that and see what the difference is afterwards.

In [81]:
world_happiness['country'] = world_happiness['country'].str.strip('*')
world_happiness['country']

RANK
1          Finland
2          Denmark
3          Iceland
4      Switzerland
5      Netherlands
          ...     
142       Botswana
143         Rwanda
144       Zimbabwe
145        Lebanon
146    Afghanistan
Name: country, Length: 146, dtype: object

In [83]:
# Difference between reference and world happiness
w_h_countries = set(world_happiness['country'].values)
print(reference_countries.difference(w_h_countries))
print(w_h_countries.difference(reference_countries))

{'Taiwan', 'Qatar', 'Maldives', 'Guyana', 'Somalia', 'Hong Kong', 'Oman', 'Angola', 'Bahamas', 'Brunei', 'Isle Of Man', 'Syria', 'Sudan', 'Czech Republic', 'Papua New Guinea', 'Bosnia And Herzegovina', 'Palestine', 'Cuba', 'Belize', 'Trinidad And Tobago', 'Fiji', 'Puerto Rico'}
{'Mozambique', 'Guinea', 'Sierra Leone', 'Chad', 'Czechia', 'Bosnia and Herzegovina', 'Laos', 'Congo', 'Lesotho', 'Turkmenistan', 'Togo', 'Benin', 'Comoros', 'Yemen', 'Hong Kong S.A.R. of China', 'Niger', 'Senegal', 'Gambia', 'Kosovo', 'Mali', 'Ivory Coast', 'Malawi', 'Mauritania', 'Tajikistan', 'Gabon', 'Taiwan Province of China', 'Liberia', 'Burkina Faso', 'Eswatini, Kingdom of', 'North Cyprus', 'Madagascar', 'Palestinian Territories'}


I can immediately see countries with the same name but different data set namings. I will create a dictionary in order to replace them.

There are 22 country names diffrent than in the world happiness country, which would bring us to 114. That is a fair number. I will manually check which ones only have the wrong namings and replace them.

In [87]:
country_namings_w_h = {'Hong Kong S.A.R. of China': 'Hong Kong', 'Taiwan Province of China': 'Taiwan', 
                   'Czechia': 'Czech Republic', 'Guinea': 'Papua New Guinea', 
                   'Bosnia and Herzegovina': 'Bosnia And Herzegovina', 'Palestinian Territories': 'Palestine'}
world_happiness.loc[
    world_happiness['country'].isin(country_namings_w_h.keys()), 'country'] = world_happiness.loc[
    world_happiness['country'].isin(country_namings_w_h.keys()), 'country'].apply(lambda x: country_namings_w_h.get(x))

In [88]:
# Difference between reference and world happiness
w_h_countries = set(world_happiness['country'].values)
print(reference_countries.difference(w_h_countries))
print(w_h_countries.difference(reference_countries))

{'Belize', 'Bahamas', 'Somalia', 'Brunei', 'Isle Of Man', 'Qatar', 'Syria', 'Sudan', 'Oman', 'Trinidad And Tobago', 'Fiji', 'Maldives', 'Guyana', 'Cuba', 'Puerto Rico', 'Angola'}
{'Mozambique', 'Sierra Leone', 'Chad', 'Laos', 'Congo', 'Lesotho', 'Turkmenistan', 'Togo', 'Benin', 'Comoros', 'Yemen', 'Niger', 'Senegal', 'Gambia', 'Kosovo', 'Mali', 'Ivory Coast', 'Malawi', 'Mauritania', 'Tajikistan', 'Gabon', 'Liberia', 'Burkina Faso', 'Eswatini, Kingdom of', 'North Cyprus', 'Madagascar'}


In [91]:
# Update reference countries with intersection
reference_countries = reference_countries.intersection(w_h_countries)
len(reference_countries)

120

In [94]:
# Difference between reference and economy stats
e_s_countries = set(economy_stats['country_name'].values)
print(sorted(reference_countries.difference(e_s_countries)))
print(sorted(e_s_countries.difference(reference_countries)))

['Bosnia And Herzegovina', 'Egypt', 'Hong Kong', 'Iran', 'Kyrgyzstan', 'Palestine', 'Russia', 'Slovakia', 'South Korea', 'Taiwan', 'Venezuela']
['American Samoa', 'Angola', 'Antigua and Barbuda', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Egypt, Arab Rep.', 'Equatorial Guinea', 'Eswatini', 'Fiji', 'Gabon', 'Gambia, The', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Hong Kong SAR, China', 'Iran, Islamic Rep.', 'Kiribati', 'Korea, Rep.', 'Kosovo', 'Kyrgyz Republic', 'Lao PDR', 'Lesotho', 'Liberia', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Marshall Islands', 'Mauritania', 'Micronesia, Fed. Sts.', 'Monaco', 'Mozambique', 'Nauru', 'Niger', 'Oman', 'Palau', 'Puerto Rico', 'Qatar

Again we can observe different names for same countries

In [96]:
country_namings_e_s = {'Hong Kong SAR, China': 'Hong Kong', 'Iran, Islamic Rep.': 'Iran', 'Czechia': 'Czech Republic',
                       'Slovak Republic': 'Slovakia', 'Bosnia and Herzegovina': 'Bosnia And Herzegovina',
                       'Korea, Rep.': 'South Korea', 'Kyrgyz Republic': 'Kyrgyzstan', 'Russian Federation': 'Russia',
                       'Egypt, Arab Rep.': 'Egypt'}
economy_stats.loc[
    economy_stats['country_name'].isin(country_namings_e_s.keys()), 'country_name'] = economy_stats.loc[
    economy_stats['country_name'].isin(country_namings_e_s.keys()), 'country_name'].apply(
    lambda x: country_namings_e_s.get(x))

In [97]:
# Difference between reference and economy stats
e_s_countries = set(economy_stats['country_name'].values)
print(sorted(reference_countries.difference(e_s_countries)))
print(sorted(e_s_countries.difference(reference_countries)))

['Palestine', 'Taiwan', 'Venezuela']
['American Samoa', 'Angola', 'Antigua and Barbuda', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Equatorial Guinea', 'Eswatini', 'Fiji', 'Gabon', 'Gambia, The', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Kiribati', 'Kosovo', 'Lao PDR', 'Lesotho', 'Liberia', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Marshall Islands', 'Mauritania', 'Micronesia, Fed. Sts.', 'Monaco', 'Mozambique', 'Nauru', 'Niger', 'Oman', 'Palau', 'Puerto Rico', 'Qatar', 'Samoa', 'Sao Tome and Principe', 'Senegal', 'Seychelles', 'Sierra Leone', 'Solomon Islands', 'Somalia', 'South Sudan', 'St. Kitts and Nevis', 'St. Lucia', 'St. Vincent and the Grenadines', 'Sudan', 'Suriname', 'Tajikistan', 'Timo

In [98]:
# We lose 3 more countries. We go on.
# Update reference countries with intersection
reference_countries = reference_countries.intersection(e_s_countries)
len(reference_countries)

117

In [99]:
# Difference between reference and financial indicators
f_i_countries = set(financial_indicators['country_name'].values)
print(sorted(reference_countries.difference(f_i_countries)))
print(sorted(f_i_countries.difference(reference_countries)))

['Afghanistan', 'Argentina', 'Bosnia And Herzegovina', 'Egypt', 'Ethiopia', 'Hong Kong', 'Iran', 'Kyrgyzstan', 'Libya', 'Myanmar', 'Russia', 'Slovakia', 'South Korea', 'Uzbekistan']
['Antigua and Barbuda', 'Bahamas, The', 'Barbados', 'Benin', 'Bhutan', 'Bosnia and Herzegovina', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Congo, Rep.', "Cote d'Ivoire", 'Djibouti', 'Dominica', 'Egypt, Arab Rep.', 'Equatorial Guinea', 'Fiji', 'Gabon', 'Gambia, The', 'Grenada', 'Guinea', 'Guyana', 'Haiti', 'Hong Kong SAR, China', 'Iran, Islamic Rep.', 'Korea, Rep.', 'Kosovo', 'Kyrgyz Republic', 'Lao PDR', 'Lesotho', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Mauritania', 'Mozambique', 'Niger', 'Oman', 'Palau', 'Qatar', 'Russian Federation', 'Samoa', 'Senegal', 'Seychelles', 'Sierra Leone', 'Slovak Republic', 'Solomon Islands', 'South Sudan', 'St. Kitts and Nevis', 'St. Lucia', 'St. Vincent and the Grenadines', 'S

In [101]:
country_namings_f_i = {'Hong Kong SAR, China': 'Hong Kong', 'Iran, Islamic Rep.': 'Iran',
                       'Slovak Republic': 'Slovakia', 'Bosnia and Herzegovina': 'Bosnia And Herzegovina',
                       'Korea, Rep.': 'South Korea', 'Kyrgyz Republic': 'Kyrgyzstan', 'Russian Federation': 'Russia',
                       'Egypt, Arab Rep.': 'Egypt'}
financial_indicators.loc[
    financial_indicators['country_name'].isin(country_namings_f_i.keys()), 'country_name'] = financial_indicators.loc[
    financial_indicators['country_name'].isin(country_namings_f_i.keys()), 'country_name'].apply(
    lambda x: country_namings_f_i.get(x))

In [102]:
# Difference between reference and financial indicators
f_i_countries = set(financial_indicators['country_name'].values)
print(sorted(reference_countries.difference(f_i_countries)))
print(sorted(f_i_countries.difference(reference_countries)))

['Afghanistan', 'Argentina', 'Ethiopia', 'Libya', 'Myanmar', 'Uzbekistan']
['Antigua and Barbuda', 'Bahamas, The', 'Barbados', 'Benin', 'Bhutan', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Congo, Rep.', "Cote d'Ivoire", 'Djibouti', 'Dominica', 'Equatorial Guinea', 'Fiji', 'Gabon', 'Gambia, The', 'Grenada', 'Guinea', 'Guyana', 'Haiti', 'Kosovo', 'Lao PDR', 'Lesotho', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Mauritania', 'Mozambique', 'Niger', 'Oman', 'Palau', 'Qatar', 'Samoa', 'Senegal', 'Seychelles', 'Sierra Leone', 'Solomon Islands', 'South Sudan', 'St. Kitts and Nevis', 'St. Lucia', 'St. Vincent and the Grenadines', 'Sudan', 'Suriname', 'Togo', 'Tonga', 'Trinidad and Tobago', 'West Bank and Gaza']


I am not ready to leave countries like Argentina and Uzbekistan outside of this analysis. I will check the other datasets and see what other countries are miissing. Perhaps we can use other countries' statistics to impute values for these 6.

In [100]:
# Difference between reference and health stats
h_s_countries = set(health_stats['country_name'].values)
print(sorted(reference_countries.difference(h_s_countries)))
print(sorted(h_s_countries.difference(reference_countries)))

['Bosnia And Herzegovina', 'Egypt', 'Hong Kong', 'Iran', 'Kyrgyzstan', 'Russia', 'Slovakia', 'South Korea']
['American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Aruba', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'British Virgin Islands', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Channel Islands', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Egypt, Arab Rep.', 'Equatorial Guinea', 'Eswatini', 'Faroe Islands', 'Fiji', 'French Polynesia', 'Gabon', 'Gambia, The', 'Gibraltar', 'Greenland', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Hong Kong SAR, China', 'Iran, Islamic Rep.', 'Isle of Man', 'Kiribati', "Korea, Dem. People's Rep.", 'Korea, Rep.', 'Kosovo', 'Kyrgyz Republic', 'Lao PDR', 'Lesotho', 'Liberia', 'Liechtenstein', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldi

In [103]:
country_namings_h_s = {'Hong Kong SAR, China': 'Hong Kong', 'Iran, Islamic Rep.': 'Iran',
                       'Slovak Republic': 'Slovakia', 'Bosnia and Herzegovina': 'Bosnia And Herzegovina',
                       'Korea, Rep.': 'South Korea', 'Kyrgyz Republic': 'Kyrgyzstan', 'Russian Federation': 'Russia',
                       'Egypt, Arab Rep.': 'Egypt'}
health_stats.loc[
    health_stats['country_name'].isin(country_namings_h_s.keys()), 'country_name'] = health_stats.loc[
    health_stats['country_name'].isin(country_namings_h_s.keys()), 'country_name'].apply(
    lambda x: country_namings_h_s.get(x))

In [104]:
# Difference between reference and health stats
h_s_countries = set(health_stats['country_name'].values)
print(sorted(reference_countries.difference(h_s_countries)))
print(sorted(h_s_countries.difference(reference_countries)))

[]
['American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Aruba', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'British Virgin Islands', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Channel Islands', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Equatorial Guinea', 'Eswatini', 'Faroe Islands', 'Fiji', 'French Polynesia', 'Gabon', 'Gambia, The', 'Gibraltar', 'Greenland', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Isle of Man', 'Kiribati', "Korea, Dem. People's Rep.", 'Kosovo', 'Lao PDR', 'Lesotho', 'Liberia', 'Liechtenstein', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Marshall Islands', 'Mauritania', 'Micronesia, Fed. Sts.', 'Monaco', 'Mozambique', 'Nauru', 'New Caledonia', 'Niger', 'Northern Mariana Islands', 'Oman', 'Palau', 'Puerto Rico', 'Qatar', 'Samoa', 'San Marino', 'Sao To

In [105]:
# Difference between reference and population and environment stats
p_e_countries = set(population_and_environment_stats['country_name'].values)
print(sorted(reference_countries.difference(p_e_countries)))
print(sorted(p_e_countries.difference(reference_countries)))

['Bosnia And Herzegovina', 'Egypt', 'Hong Kong', 'Iran', 'Kyrgyzstan', 'Russia', 'Slovakia', 'South Korea']
['American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Aruba', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'British Virgin Islands', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Channel Islands', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Egypt, Arab Rep.', 'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Faroe Islands', 'Fiji', 'French Polynesia', 'Gabon', 'Gambia, The', 'Gibraltar', 'Greenland', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Hong Kong SAR, China', 'Iran, Islamic Rep.', 'Isle of Man', 'Kiribati', "Korea, Dem. People's Rep.", 'Korea, Rep.', 'Kyrgyz Republic', 'Lao PDR', 'Lesotho', 'Liberia', 'Liechtenstein', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Mald

In [106]:
country_namings_p_e = {'Hong Kong SAR, China': 'Hong Kong', 'Iran, Islamic Rep.': 'Iran',
                       'Slovak Republic': 'Slovakia', 'Bosnia and Herzegovina': 'Bosnia And Herzegovina',
                       'Korea, Rep.': 'South Korea', 'Kyrgyz Republic': 'Kyrgyzstan', 'Russian Federation': 'Russia',
                       'Egypt, Arab Rep.': 'Egypt'}
population_and_environment_stats.loc[
    population_and_environment_stats['country_name'].isin(
        country_namings_p_e.keys()), 'country_name'] = population_and_environment_stats.loc[
    population_and_environment_stats['country_name'].isin(country_namings_p_e.keys()), 'country_name'].apply(
    lambda x: country_namings_p_e.get(x))

In [107]:
# Difference between reference and population and environment stats
p_e_countries = set(population_and_environment_stats['country_name'].values)
print(sorted(reference_countries.difference(p_e_countries)))
print(sorted(p_e_countries.difference(reference_countries)))

[]
['American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Aruba', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'British Virgin Islands', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Channel Islands', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Faroe Islands', 'Fiji', 'French Polynesia', 'Gabon', 'Gambia, The', 'Gibraltar', 'Greenland', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Isle of Man', 'Kiribati', "Korea, Dem. People's Rep.", 'Lao PDR', 'Lesotho', 'Liberia', 'Liechtenstein', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Marshall Islands', 'Mauritania', 'Micronesia, Fed. Sts.', 'Monaco', 'Mozambique', 'Nauru', 'New Caledonia', 'Niger', 'Northern Mariana Islands', 'Oman', 'Palau', 'Puerto Rico', 'Qatar', 'Samoa', 'San Marino', 'Sao T

In [108]:
# Difference between reference and infrastructure stats
i_s_countries = set(infrastructure_stats['country_name'].values)
print(sorted(reference_countries.difference(i_s_countries)))
print(sorted(i_s_countries.difference(reference_countries)))

['Bosnia And Herzegovina', 'Egypt', 'Hong Kong', 'Iran', 'Kyrgyzstan', 'Russia', 'Slovakia', 'South Korea']
['American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Aruba', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'British Virgin Islands', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Egypt, Arab Rep.', 'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Faroe Islands', 'Fiji', 'French Polynesia', 'Gabon', 'Gambia, The', 'Gibraltar', 'Greenland', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Hong Kong SAR, China', 'Iran, Islamic Rep.', 'Isle of Man', 'Kiribati', "Korea, Dem. People's Rep.", 'Korea, Rep.', 'Kyrgyz Republic', 'Lao PDR', 'Lesotho', 'Liberia', 'Liechtenstein', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Mar

In [109]:
country_namings_i_s = {'Hong Kong SAR, China': 'Hong Kong', 'Iran, Islamic Rep.': 'Iran',
                       'Slovak Republic': 'Slovakia', 'Bosnia and Herzegovina': 'Bosnia And Herzegovina',
                       'Korea, Rep.': 'South Korea', 'Kyrgyz Republic': 'Kyrgyzstan', 'Russian Federation': 'Russia',
                       'Egypt, Arab Rep.': 'Egypt'}
infrastructure_stats.loc[
    infrastructure_stats['country_name'].isin(
        country_namings_i_s.keys()), 'country_name'] = infrastructure_stats.loc[
    infrastructure_stats['country_name'].isin(country_namings_i_s.keys()), 'country_name'].apply(
    lambda x: country_namings_i_s.get(x))

In [110]:
# Difference between reference and infrastructure stats
i_s_countries = set(infrastructure_stats['country_name'].values)
print(sorted(reference_countries.difference(i_s_countries)))
print(sorted(i_s_countries.difference(reference_countries)))

[]
['American Samoa', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Aruba', 'Bahamas, The', 'Barbados', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'British Virgin Islands', 'Brunei Darussalam', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', "Cote d'Ivoire", 'Cuba', 'Curacao', 'Djibouti', 'Dominica', 'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Faroe Islands', 'Fiji', 'French Polynesia', 'Gabon', 'Gambia, The', 'Gibraltar', 'Greenland', 'Grenada', 'Guam', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Isle of Man', 'Kiribati', "Korea, Dem. People's Rep.", 'Lao PDR', 'Lesotho', 'Liberia', 'Liechtenstein', 'Macao SAR, China', 'Madagascar', 'Malawi', 'Maldives', 'Mali', 'Marshall Islands', 'Mauritania', 'Micronesia, Fed. Sts.', 'Monaco', 'Mozambique', 'Nauru', 'New Caledonia', 'Niger', 'Northern Mariana Islands', 'Oman', 'Palau', 'Puerto Rico', 'Qatar', 'Samoa', 'San Marino', 'Sao Tome and Principe', 

It looks like for all the other data sets we have the values for all the 117 countries. I will attempt to fill out the financial indicators for the 6 countries individually.

In [115]:
# Replace values individually
financial_indicators.loc[300, :] = ['Argentina', 6, 3]
#financial_indicators.loc[financial_indicators['country_name'].isin(reference_countries.difference(f_i_countries))]
financial_indicators.tail()

series_name        country_name  Consumer price index (2010 = 100)  \
160                     Vietnam                           0.996434   
161          West Bank and Gaza                           0.999327   
162                      Zambia                           0.992801   
163                    Zimbabwe                           0.870794   
300                   Argentina                           6.000000   

series_name  Inflation, consumer prices (annual %)  
160                                       0.993002  
161                                       0.994069  
162                                       0.968152  
163                                       0.000000  
300                                       3.000000

I cannot simply replace the values here, sicne I already preprocessed them. Therefore, for this particular dataset I will reiterate the preprocessing steps, but add the 6 needed rows for ['Afghanistan', 'Argentina', 'Ethiopia', 'Libya', 'Myanmar', 'Uzbekistan'] first